### ISSUE TITLE
intercept and modify queries on the fly at high level

### ISSUE THREAD

**Issue by MohammadMQ** [dotnet/efcore#12199]
Hi I have a comment regarding the use of Interceptions.
Currently there is no really support for intercepting read queries like select, where, distinct etc. just as with create, update and delete using override SaveChanges. 
There is however the possibility to use diagnostics like this 

`        [DiagnosticName("Microsoft.EntityFrameworkCore.Database.Command.CommandExecuting")]
        public void OnCommandExecuting(DbCommand command, DbCommandMethod executeMethod, Guid commandId, Guid connectionId, bool async, DateTimeOffset startTime)
        {`
or 
`        [DiagnosticName("Microsoft.EntityFrameworkCore.Database.Command.CommandExecuted")]
        public void OnCommandExecuted(object result, bool async)
        {`

but this is very low level. 
What I need is the ability to hook into the returned query result set (in order to see what customerId is inside each row, so that I may compare it to currently logged in user and thereby do some security checking). 
In short, I a need some high lewel interception (just like with override SaveChanges).

Currently I'm using PostSharp (a commercial product) to do this, but it would be great if EF Core would also provide this facility . 

### Further technical details
EF Core version: 2.0
Database Provider: PostGreSql
Operating system: Windows 10 pro
IDE: Visual Studio 2017 15.7.2

**Comment by divega** [comment]
@MohammadQueue-it from the description it is not clear to me if you really need what we usually mean by interception. "Intercepting and modifying" the query usually implies manipulating (i.e. rewriting) the query before it is executed. But you say you are actually looking into the verifying something on the query results, which seems to imply that you want a hook that it is going to allow you to look at the results after the query has been executed. 

This is just connecting the dots but if what you want is to make sure the results only include data that belongs to the current user, you may be better off using [global query filters](https://docs.microsoft.com/en-us/ef/core/querying/filters). If you instead really want to look into entities entering the DbContext as a result of query execution, you may want to take advantage of the `Tracked` event that is new in 2.1.

**Comment by MohammadQueue-it** [comment]
Hi 
Thanks for your reply
What I want is exactly what you mentioned: "...want a hook that it is going to allow you to look at the results after the query has been executed."
I have looked at Global Query Filters, however I cannot see how I can achieve that using e.g. OnModelCreating, modelbuilder. However I will take a look at the new Tracked event feature in 2.1 that you are referring to.

**Comment by smitpatel** [comment]
@MohammadQueue-it - With regards to query filter use in your case, you can a property on your db context which stores id of logged in user and use that property inside your query filter. When running the query set the property on db context with current user and query will use that automatically to do filtering if needed.

**Comment by queueit** [comment]
Hi smitpatel
I don't want to add security layer inside dbContext

**Comment by weitzhandler** [comment]
@divega Thanks for your comment.
Is there a way to intercept lower level database actions, such as the database generation process itself, or at least a notification upon its occurrence?
